exec/cnex/lib/net.c

```diff
@@ -0,0 +1,310 @@
+#include "net.h"
+
+#include <errno.h>
+#include <iso646.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "array.h"
+#include "cell.h"
+#include "exec.h"
+#include "socketx.h"
+#include "stack.h"
+#include "nstring.h"
+
+
+
+#ifdef _WIN32
+static void initWinsock()
+{
+    static BOOL initialized = FALSE;
+    if (!initialized) {
+        WSADATA wsa;
+        WSAStartup(MAKEWORD(1, 1), &wsa);
+        initialized = TRUE;
+    }
+}
+#endif
+
+void net_tcpSocket(TExecutor *exec)
+{
+#ifdef _WIN32
+    initWinsock();
+#endif
+    Cell *r = cell_createArrayCell(0);
+    SOCKET s = socket(AF_INET, SOCK_STREAM, 0);
+    Cell handle; handle.type = cNumber;
+    handle.number = number_from_sint32((int32_t)(s));
+    cell_arrayAppendElement(r, handle);
+    push(exec->stack, r);
+}
+
+void net_udpSocket(TExecutor *exec)
+{
+#ifdef _WIN32
+    initWinsock();
+#endif
+    Cell *r = cell_createArrayCell(0);
+    SOCKET s = socket(AF_INET, SOCK_DGRAM, 0);
+    Cell handle; handle.type = cNumber;
+    handle.number = number_from_sint32((int32_t)(s));
+    cell_arrayAppendElement(r, handle);
+    push(exec->stack, r);
+}
+
+void net_socket_accept(TExecutor *exec)
+{
+    SOCKET s = number_to_sint32(top(exec->stack)->number); pop(exec->stack);
+
+    struct sockaddr_in sin;
+    socklen_t slen = sizeof(sin);
+    SOCKET r = accept(s, (struct sockaddr *)(&sin), &slen);
+    if (r < 0) {
+        perror("accept");
+        push(exec->stack, cell_newCell());
+        return;
+    }
+    Cell *client = cell_createArrayCell(0);
+    Cell handle; handle.type = cNumber;
+    handle.number = number_from_sint32((int32_t)(r));
+    cell_arrayAppendElement(client, handle);
+
+    push(exec->stack, client);
+}
+
+void net_socket_bind(TExecutor *exec)
+{
+    Number port = peek(exec->stack, 0)->number;
+    const char *address = string_ensureNullTerminated(peek(exec->stack, 1)->string);
+    Cell *handle = peek(exec->stack, 2);
+
+    SOCKET s = number_to_sint32(handle->number);
+    int p = number_to_sint32(port);
+    struct in_addr addr;
+    if (strlen(address) == 0) {
+        addr.s_addr = INADDR_ANY;
+    } else {
+        addr.s_addr = inet_addr(address);
+        if (addr.s_addr == INADDR_NONE) {
+            struct hostent *he = gethostbyname(address);
+            if (he == NULL) {
+                // ToDo: Raise exception here, that host does not exist?
+                goto bail;
+            }
+            addr = *(struct in_addr*)he->h_addr_list;
```

Comment:
This probably won't work, you're not using `h_addr_list` correctly.

Reply:
Fixed.